server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java

```diff
@@ -0,0 +1,282 @@
+        }
+
+        public int compareWith(Category that) {
+            return order - that.order;
+        }
+
+    }
+
+
+    private CredentialTypeMetadata() {
+    }
+
+
+    // GETTERS
+
+    /**
+     * @return credential type like for example "password", "otp" or "webauthn"
+     */
+    public String getType() {
+        return type;
+    }
+
+   /**
+     * @return the label, which will be shown to the end user on various screens, like login screen with available authentication mechanisms.
+     * This label will reference this particular authenticator type.
+     * It should be clear to end users. For example, implementations can return "Authenticator Application" for OTP or "Security Key" for WebAuthn.
+     *
+     * Alternatively, this method can return a message key, so that it is possible to localize it for various languages.
+     */
+    public String getDisplayName() {
+        return displayName;
+    }
+
+    /**
+     * @return the text, which will be shown to the user on various screens, like login screen with available authentication mechanisms.
+     * This text will reference this particular authenticator type.
+     * For example for OTP, the returned text could be "Enter a verification code from authenticator application" .
+     *
+     * Alternatively, this method can return a message key, so that it is possible to localize it for various languages.
+     */
+    public String getHelpText() {
+        return helpText;
+    }
+
+    /**
+     * Return the icon CSS, which can be used to display icon, which represents this particular authenticator.
+     *
+     * The icon will be displayed on various places. For example the "Select authenticator" screen during login, where user can select from
+     * various authentication mechanisms for two-factor or passwordless authentication.
+     *
+     * The returned value can be either:
+     * - Key of the property, which will reference the actual CSS in the themes.properties file. For example if you return "kcAuthenticatorWebAuthnClass"
+     *   from this method, then your themes.properties should have the property like for example "kcAuthenticatorWebAuthnClass=fa fa-key list-view-pf-icon-lg" .
+     *   This would mean that "fa fa-key list-view-pf-icon-lg" will be the actual CSS used.
+     * - the icon CSS class directly. For example you can return "fa fa-key list-view-pf-icon-lg" directly for the above example with WebAuthn.
+     *   This alternative is fine just if your authenticator can use same CSS class for all the themes.
+     *
+     * If you don't expect your authenticator to need icon (for example it will never be shown in the "select authenticator" screen), then
+     * it is fine to keep the default value.
+     */
+    public String getIconCssClass() {
+        return iconCssClass;
+    }
+
+    /**
+     * @return the providerID of the required action, which can be used by the user to create new credential of our type. Null if there is no
```

Comment:
Should this be providerID?  For instance, the create action for OTP is "CONFIGURE_TOTP".

Reply:
Yes, "CONFIGURE_TOTP" is the providerID of the required action provider. When requiredAction/AIA is triggered, Keycloak uses the providerID like "CONFIGURE_TOTP" to lookup required action provider - in this case class UpdateTotp. In case of WebAuthn, the name of the action is "webauthn-register" , which is provider id of WebAuthnRegisterFactory.

By default, the alias and providerId of RequiredActionProviderModels attached to the realm is the same, but it may not be necessarily true. Keycloak uses providerId to lookup the requiredAction, not the alias or name of requiredAction.

However if you have any better idea to this javadoc, we can discuss it.

Reply:
This is fine.  I just didn't remember it being the providerId.  I was thinking the action name was separate.